File: app/src/main/java/me/tagavari/airmessage/util/CompoundErrorDetails.java
package me.tagavari.airmessage.util;

import androidx.annotation.Nullable;

import me.tagavari.airmessage.connection.exception.AMRequestException;
import me.tagavari.airmessage.enums.ChatCreateErrorCode;
import me.tagavari.airmessage.enums.MessageSendErrorCode;

/**
 * Represents error information that has an error code and an optional error detail string
 */
public abstract class CompoundErrorDetails {
	final int error;
	final String detail;
	
	/**
	 * Represents an error for a message
	 * @param error The error code
	 * @param detail The error detail string, if available
	 */
	public CompoundErrorDetails(int error, @Nullable String detail) {
		this.error = error;
		this.detail = detail;
	}
	
	public int getError() {
		return error;
	}
	
	@Nullable
	public String getDetail() {
		return detail;
	}
	
	public AMRequestException toException() {
		return new AMRequestException(error, detail);
	}
	
	public static class MessageSend extends CompoundErrorDetails {
		public MessageSend(@MessageSendErrorCode int error, @Nullable String detail) {
			super(error, detail);
		}
	}
	
	public static class ChatCreate extends CompoundErrorDetails {
		public ChatCreate(@ChatCreateErrorCode int error, @Nullable String detail) {
			super(error, detail);
		}
	}
}

File: app/src/main/java/me/tagavari/airmessage/util/CompoundErrorDetails.kt
package me.tagavari.airmessage.util

import me.tagavari.airmessage.connection.exception.AMRequestException
import me.tagavari.airmessage.enums.ChatCreateErrorCode
import me.tagavari.airmessage.enums.MessageSendErrorCode

/**
 * Represents error information that has an error code and an optional error detail string
 * @param error The error code
 * @param detail The error detail string, if available
 */
abstract class CompoundErrorDetails(val error: Int, val detail: String?) {
	fun toException(): AMRequestException {
		return AMRequestException(error, detail)
	}
	
	class MessageSend(@MessageSendErrorCode error: Int, detail: String?) : CompoundErrorDetails(error, detail)
	class ChatCreate(@ChatCreateErrorCode error: Int, detail: String?) : CompoundErrorDetails(error, detail)
}